llvm/lib/Target/SystemZ/SystemZISelLowering.cpp

```diff
@@ -8702,61 +8738,264 @@ SDValue SystemZTargetLowering::combineSETCC(
+  if (CCNode->getOpcode() == SystemZISD::TM) {
+    if (CCValid != SystemZ::CCMASK_TM)
       return false;
-
-    // Verify that the ICMP compares against one of select values.
-    auto *TrueVal = dyn_cast<ConstantSDNode>(CompareLHS->getOperand(0));
-    if (!TrueVal)
+    if ((CCMask != SystemZ::CCMASK_TM_SOME_1) &&
+        (CCMask != SystemZ::CCMASK_TM_ALL_0))
       return false;
-    auto *FalseVal = dyn_cast<ConstantSDNode>(CompareLHS->getOperand(1));
-    if (!FalseVal)
+    SDValue Op0 = CCNode->getOperand(0);
+    SDValue Op1 = CCNode->getOperand(1);
+    auto [Op0CC, Op0CCValid] = findCCUse(Op0);
+    if (Op0CC == SDValue())
       return false;
-    if (CompareRHS->getAPIntValue() == FalseVal->getAPIntValue())
-      Invert = !Invert;
-    else if (CompareRHS->getAPIntValue() != TrueVal->getAPIntValue())
+    int TMConstVal;
+    auto *N = dyn_cast<ConstantSDNode>(Op1);
+    if (N) {
+      TMConstVal = N->getZExtValue();
+      if ((TMConstVal != (1 << SystemZ::IPM_CC)) && (TMConstVal != 1))
+        return false;
+    }
+    auto &&Op0CCVals = simplifyAssumingCCVal(Op0, Op0CC);
+    auto [Op1CC, Op1CCValid] = findCCUse(Op1);
+    auto Op1CCVals =
+        simplifyAssumingCCVal(Op1, Op1CC == SDValue() ? Op0CC : Op1CC);
+    if (Op0CCVals.empty() || Op1CCVals.empty())
       return false;
-
-    // Compute the effective CC mask for the new branch or select.
-    auto *NewCCValid = dyn_cast<ConstantSDNode>(CompareLHS->getOperand(2));
-    auto *NewCCMask = dyn_cast<ConstantSDNode>(CompareLHS->getOperand(3));
-    if (!NewCCValid || !NewCCMask)
+    std::transform(Op0CCVals.begin(), Op0CCVals.end(), Op1CCVals.begin(),
+                   Op0CCVals.begin(), [&TMConstVal](auto &CC0, auto &CC1) {
+                     return (TMConstVal == (1 << SystemZ::IPM_CC))
+                                ? ((CC0 & CC1) >> SystemZ::IPM_CC)
+                                : (CC0 & CC1);
+                   });
+    if (std::any_of(Op0CCVals.begin(), Op0CCVals.end(),
+                    [](const auto &CC) { return CC < 0 || CC > 3; }))
       return false;
-    CCValid = NewCCValid->getZExtValue();
-    CCMask = NewCCMask->getZExtValue();
-    if (Invert)
-      CCMask ^= CCValid;
+    auto CCMaskValsSet = convertCCMaskToCCValsSet(CCMask);
+    auto CCValidValsSet = convertCCMaskToCCValsSet(CCValid);
+    CCMask = 0;
+    size_t ValidCCVal = 0; // Maybe not needed for TM.
+    for (auto CC : Op0CCVals) {
+      CCMask <<= 1;
+      CCMask |= (CCMaskValsSet.count(CC) && CCValidValsSet.count(ValidCCVal));
+      ++ValidCCVal;
+    }
+    CCReg = Op0CC;
+    return true;
+  }
+  if (CCNode->getOpcode() != SystemZISD::ICMP ||
+      CCValid != SystemZ::CCMASK_ICMP)
+    return false;
 
-    // Return the updated CCReg link.
-    CCReg = CompareLHS->getOperand(4);
+  SDValue CmpOp0 = CCNode->getOperand(0);
+  SDValue CmpOp1 = CCNode->getOperand(1);
+  if (isa<ConstantSDNode>(CmpOp0) && isa<ConstantSDNode>(CmpOp1))
+    return false;
```

Comment:
Huh, does this already happen?  If so, in this case we can simply evaluate SELECT_CCMASK fully just here ...